vic/drivers/cesm/cpl_mct/lnd_comp_mct.F90

```diff
@@ -0,0 +1,722 @@
+    TYPE(seq_cdata)             , INTENT(inout) :: cdata_s
+    TYPE(mct_aVect)             , INTENT(inout) :: x2s, s2x
+    CHARACTER(len=*), OPTIONAL  , INTENT(in)    :: NLFilename
+
+    !--- Local Variables
+    INTEGER                                     :: LNDID
+    INTEGER                                     :: mpicom_lnd
+    INTEGER                                     :: mytask, ierr
+    INTEGER                                     :: lsize, gsize
+    INTEGER                                     :: nx_global, ny_global
+    INTEGER                                     :: dtime  ! timestep (seconds)
+    INTEGER, POINTER                            :: idata(:)
+    INTEGER, DIMENSION(:), ALLOCATABLE          :: gindex  ! Number the local grid points
+    REAL(r8), POINTER                           :: lon_data(:)
+    REAL(r8), POINTER                           :: lat_data(:)
+    REAL(r8), POINTER                           :: area_data(:)
+    REAL(r8), POINTER                           :: mask_data(:)
+    REAL(r8), POINTER                           :: frac_data(:)
+    TYPE(mct_gsMap),  POINTER                   :: GSMap_lnd
+    TYPE(mct_gGrid),  POINTER                   :: dom_lnd
+    TYPE(seq_infodata_type), POINTER            :: infodata
+    CHARACTER(len=VICMAXSTRING)                 :: caseid
+    CHARACTER(len=VICMAXSTRING)                 :: starttype
+    CHARACTER(len=VICMAXSTRING)                 :: calendar
+    LOGICAL                                     :: exists  ! true if file exists
+    CHARACTER(len=*), PARAMETER                 :: subname = 'lnd_init_mct'
+    CHARACTER(len=*), PARAMETER                 :: FORMAT = "('("//TRIM(subname)//") :',A)"
+
+    !--- Local Variables for C/Fortran Interface
+    INTEGER(C_INT)                              :: errno
+    CHARACTER(len=VICMAXSTRING, kind=C_CHAR)    :: vic_global_param_file
+
+    !--- initialize the field index values, subroutine is below in this file
+    CALL cpl_indices_set()
+
+    !--- point/get data from cdata datatype
+    CALL seq_cdata_setptrs(cdata, ID=LNDID, mpicom=mpicom_lnd, &
+         gsMap=GSMap_lnd, dom=dom_lnd, infodata=infodata)
+    CALL MPI_COMM_RANK(mpicom_lnd, mytask, ierr)
+
+    !--- copy/hand the mpicom from the driver to the vic model
+    errno = vic_cesm_init_mpi(mpicom_lnd)
+    IF (errno /= 0) THEN
+       CALL shr_sys_abort(subname//':: vic_cesm_init_mpi returned a errno /= 0')
+    ENDIF
+
+    !--- get unit number for vic log file
+    !--- setup vic log file
+    !--- set shr unit number to vic log file
+    CALL shr_file_getLogUnit(shrlogunit)
+    IF (mpicom_lnd == 0) THEN
+       ! Tony look here INQUIRE(file='lnd_modelio.nml'//TRIM(inst_suffix), exist=exists)
+       exists = .FALSE.
+       IF (exists) THEN
+          iulog = shr_file_getUnit()
+          !Tony look here CALL shr_file_setIO('lnd_modelio.nml'//TRIM(inst_suffix), iulog)
+          CALL shr_file_setIO('lnd_modelio.nml', iulog)
+       ENDIF
```

Comment:
The driver and other components manage their own log files.  In some ways, vic needs to decide how to handle this.  In general, in CESM, a component should write from the root processor to the *.log file associated with lnd_modelio.nml and other tasks in that component can write to unit number 6 (which is the ccsm.log file usually).  If VIC does not want to follow that convention, that's fine.  In this case, I propose the vic driver (lnd_comp_mct) and other vic coupling layer methods use this approach and that vic "internal" write to it's own files.   By the way, in this code section, the logic should not be "IF (mpicom_lnd == 0) THEN", it should be something like
iulog = 6
if (my_task == 0) then
          iulog = shr_file_getUnit()
          CALL shr_file_setIO('lnd_modelio.nml', iulog)
endif

In other words, all tasks will write to unit 6 through the iulog unit number except the root task.  The root task will "get" a unique unit number and then assign that unit number of the lnd.log file.

One other thing that we could think about is managing the unit numbers in vic.  If there are any, we could query for free unit numbers in the init call in lnd_comp_mct and assign them to vic internal variables.  But this might not be necessary with C, not sure.